examples/api/lib/material/input_chip/input_chip.1.dart

```diff
@@ -0,0 +1,367 @@
+    required this.onChanged,
+    this.onChipTapped,
+    this.onSubmitted,
+    this.onTextChanged,
+  });
+
+  final List<T> values;
+  final InputDecoration decoration;
+  final TextStyle? style;
+  final StrutStyle? strutStyle;
+
+  final ValueChanged<List<T>> onChanged;
+  final ValueChanged<T>? onChipTapped;
+  final ValueChanged<String>? onSubmitted;
+  final ValueChanged<String>? onTextChanged;
+
+  final Widget Function(BuildContext context, T data) chipBuilder;
+
+  @override
+  ChipsInputState<T> createState() => ChipsInputState<T>();
+}
+
+class ChipsInputState<T> extends State<ChipsInput<T>> {
+  @visibleForTesting
+  late final ChipsInputEditingController<T> controller;
+
+  String _previousText = '';
+  TextSelection? _previousSelection;
+
+  @override
+  void initState() {
+    super.initState();
+
+    controller = ChipsInputEditingController<T>(
+      <T>[...widget.values],
+      widget.chipBuilder,
+    );
+    controller.addListener(_textListener);
+  }
+
+  @override
+  void dispose() {
+    controller.removeListener(_textListener);
+    controller.dispose();
+
+    super.dispose();
+  }
+
+  void _textListener() {
+    final String currentText = controller.text;
+
+    if (_previousSelection != null) {
+      final int currentNumber = countReplacements(currentText);
+      final int previousNumber = countReplacements(_previousText);
+
+      final int cursorEnd = _previousSelection!.extentOffset;
+      final int cursorStart = _previousSelection!.baseOffset;
+
+      final List<T> values = <T>[...widget.values];
+
+      // If the current number and the previous number of replacements are different, then
+      // the user has deleted the InputChip using the keyboard. In this case, we trigger
+      // the onChanged callback. We need to be sure also that the current number of
+      // replacements is different from the input chip to avoid double-deletion.
+      if (currentNumber < previousNumber && currentNumber != values.length) {
+        if (cursorStart == cursorEnd) {
+          values.removeRange(cursorStart - 1, cursorEnd);
+        } else {
+          if (cursorStart > cursorEnd) {
+            values.removeRange(cursorEnd, cursorStart);
+          } else {
+            values.removeRange(cursorStart, cursorEnd);
+          }
+        }
+        widget.onChanged(values);
+      }
+    }
+
+    _previousText = currentText;
+    _previousSelection = controller.selection;
+  }
+
+  static int countReplacements(String text) {
+    return text.codeUnits
+        .where(
+            (int u) => u == ChipsInputEditingController.kObjectReplacementChar)
+        .length;
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    controller.updateValues(<T>[...widget.values]);
+
+    return TextField(
+      minLines: 1,
+      maxLines: 3,
+      textInputAction: TextInputAction.done,
+      style: widget.style,
+      strutStyle: widget.strutStyle,
+      controller: controller,
+      onChanged: (_) =>
```

Comment:
Please include parameters and types, even if they're not used per https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-anonymous-parameter-names